compiler/syntax/src/res_printer.ml

```diff
@@ -1191,13 +1171,39 @@ and print_value_description ~state value_description cmt_tbl =
           else Doc.nil);
        ])
 
-and print_type_declarations ?inline_record_definitions ~state ~rec_flag
-    type_declarations cmt_tbl =
-  print_listi
-    ~get_loc:(fun n -> n.Parsetree.ptype_loc)
-    ~nodes:type_declarations
-    ~print:(print_type_declaration2 ?inline_record_definitions ~state ~rec_flag)
-    cmt_tbl
+and print_type_declarations ~state ~rec_flag type_declarations cmt_tbl =
+  if has_inline_type_definitions type_declarations then
+    let inline_record_definitions, regular_declarations =
+      type_declarations
+      |> List.partition (fun (td : Parsetree.type_declaration) ->
+             Res_parsetree_viewer.has_inline_record_definition_attribute
+               td.ptype_attributes)
+    in
+    let adjusted_rec_flag =
+      match rec_flag with
+      | Recursive ->
+        if List.length regular_declarations > 1 then Doc.text "rec "
+        else Doc.nil
+      | Nonrecursive -> Doc.nil
+    in
+    print_listi
+      ~get_loc:(fun n -> n.Parsetree.ptype_loc)
+      ~nodes:regular_declarations
+      ~print:
+        (print_type_declaration2 ~inline_record_definitions ~state
+           ~rec_flag:adjusted_rec_flag)
+      cmt_tbl
```

Comment:
Move the inline record printing logic into `print_type_declarations` itself. That's where it belongs IMO.